### DISCUSSION TITLE
How to get subtree or subgraph before a node？

### DISCUSSION THREAD

**Discussion by forestbat** [networkx/networkx#5897] [Q&A] [answered]
For example，I write a graph like this：
(0,2),(1,2),(2,3),(3,4)
I can use dfs_tree(graph,2) to get node 3 and 4，however I don't know how to get the subtree/subgraph which contains {0,1,2}，and I don't want to use ancestors().
My true purpose is to find upstream of a node on a river.

**Comment by dschult** [comment, marked as answer]
Why don't you want to use ancestors?

**Reply by forestbat** [reply]
> Why don't you want to use ancestors?

It's only a set without any tree/graph structure,now I can only find simple paths.

**Reply by dschult** [reply]
Try something like:  `G2 = G.subgraph(nx.ancestors(G, 2)).copy()`
The `subgraph` turns the node set into the graph induced on those nodes. The `copy` is not necessary, but it separates the subgraph object "view" from the graph so manipulations are possible and subsequent pathfinding can work faster.

**Reply by forestbat** [reply]
> Try something like: `G2 = G.subgraph(nx.ancestors(G, 2)).copy()` The `subgraph` turns the node set into the graph induced on those nodes. The `copy` is not necessary, but it separates the subgraph object "view" from the graph so manipulations are possible and subsequent pathfinding can work faster.

Thanks for your reply，and how can I find all simple paths from nodes which in-degree are 0 to node which I want to research such as 2？

**Reply by dschult** [reply]
https://networkx.org/documentation/stable/reference/algorithms/simple_paths.html
